### PR TITLE
Add Promote button and modal

### PR DESCRIPTION
**Pull request by mikeyarce**
Fixes https://github.com/Automattic/WP-Job-Manager/issues/2445 https://github.com/Automattic/WP-Job-Manager/issues/2438

### Changes proposed in this Pull Request

* In the Job Listings Admin page, it moves all the Actions from buttons to links under the job titles
* Adds a new Column called "Promote" which has a button for promoting jobs
* When you click on `Promote`, it opens a dialog box which contains a shadow-dom template that's currently populated by hand.  This is just temporary until we have a way to populate it from wpjobmanager.com
* The dialog is added as an `admin-notice` just as a way to have the HTML in the DOM, that seemed the simplest way
* There's only one dialog element, but we can use JS to dynamically change the link.  Right now, as an example, the `href` attribute is set to the POST ID - we can change this to wahtever we need.

### Next tasks
- Get the HTML fragment from wpjobmanager.com 
- Store the fragment in some kind of cache
- Add proper links to the `Promote your job` and `Learn more` buttons

### Testing instructions

* Apply this PR
* Run `npm run start` to build assets
* Go to Job Listings -> All Jobs 
* Make sure the actions work (Edit, View, Approve, Delete)
* Click on "Promote" and see the modal
* Make sure you can click the `X` to leave the dialog
* Make sure you can leave the dialog by pressing `ESC`
* Set the Post Meta of a job listing to `_promoted => 1` and make sure you see the links to Edit and Deactivate the promotion (the links don't work yet)

### Screenshot / Video
Screenshot 2023-06-17 at 3 44 07 PM

promote-dialog

### CODE REVIEW

File: assets/css/admin.scss
Comment (renatho) [inline assets/css/admin.scss]: It's worth adding a class here instead of `dialog`, otherwise it could apply to external elements that we don't want.

File: assets/js/admin.js
Comment (renatho) [inline assets/js/admin.js]: I think we should also use camelCase in this variable, right?

File: assets/css/admin.scss
Comment (renatho) [inline assets/css/admin.scss]: I noticed that the styles of the "Starting from" text is not matching the design. In order to make it simpler, I'd suggest separating it into 2 classes.

File: assets/css/admin.scss
Comment (renatho) [inline assets/css/admin.scss]: I'm not sure about new code we're writing in WP Job Manager, so I'd like to ask you if we're not using [BEM](https://getbem.com/naming/) similar to Sensei.
If not, I'd suggest some more specific class here. Maybe `promote-job-modal-price`? So we avoid getting `price` class style from other plugins, since it's a very common name. The same I think also applies to the `column-left` and `column-right` classes.

File: assets/css/admin.scss
Comment (renatho) [inline assets/css/admin.scss]: The `li` seems a little more close here than in the design. Maybe we could add some extra margin?

File: assets/js/admin.js
Comment (renatho) [inline assets/js/admin.js]: Since it's written in `admin.css` and `admin.js`, and we're creating a custom element, maybe we could use generic names? So we could use it in the future for other things.

And we'd need to add only the specific Promoted Jobs logic separated (update the `href`)?

File: includes/admin/class-wp-job-manager-cpt.php
Comment (renatho) [inline includes/admin/class-wp-job-manager-cpt.php:580]: I wonder if it could be a concern for other plugins using this filter, trying to use it with the previous styles. Have you checked the other WP Job Manager extensions to see if it's not being used?

File: includes/admin/class-wp-job-manager-promoted-jobs.php
Comment (renatho) [inline includes/admin/class-wp-job-manager-promoted-jobs.php]: I think it's supposed to be `int`, right?

```suggestion
	 * @param int $post_id
```

File: includes/admin/class-wp-job-manager-promoted-jobs.php
Comment (renatho) [inline includes/admin/class-wp-job-manager-promoted-jobs.php]: Do we need this `<style>` part? (I'm not used to `template`, so please bear with me with the questions here 🤭).

File: includes/admin/class-wp-job-manager-promoted-jobs.php
Comment (renatho) [inline includes/admin/class-wp-job-manager-promoted-jobs.php]: I think it's missing a tab here. Also, maybe it's worth adding `i18n` functions here?

File: includes/admin/class-wp-job-manager-promoted-jobs.php
Comment (renatho) [inline includes/admin/class-wp-job-manager-promoted-jobs.php]: Since it's a default content that could be used in some situation, maybe we should only use more generic information?
If it makes sense, we would need to be a little creative here to write some stuff that will never change, like _"Promote your job on external job boards"_.
It would also apply to the price, maybe having something like "$ --".

File: assets/js/admin.js
Comment (renatho) [inline assets/js/admin.js]: I imagine you used spaces instead of tabs accidentally here, right?

File: includes/admin/class-wp-job-manager-promoted-jobs.php
Comment (renatho) [inline includes/admin/class-wp-job-manager-promoted-jobs.php]: I wonder if another hook more generic, like `admin_footer` would make more sense here.

Also, I imagine this class is something generic for many things related to promoted jobs, right? If yes, I'd suggest to use some more specific name for these methods, just to avoid confusion when the class grows.

**Review by renatho — COMMENTED** [review]
Hi @mikeyarce! It's looking very nice with the `dialog` and the `template`. =)

I just added a few suggestions and some comments to confirm. Also, I'd suggest to add Zen team as reviewer since I don't work too much on WP Job Manager and maybe I could miss something.

File: includes/admin/class-wp-job-manager-cpt.php
Comment (mikeyarce) [reply includes/admin/class-wp-job-manager-cpt.php:580]: I tried to copy the use from the previous implementation so that it won't be different for other plugins that previous used the filters.  What do you think could happen in this case?

Comment (mikeyarce) [reply includes/admin/class-wp-job-manager-cpt.php:580]: Also, I think we should absolutely check with third-party devs ahead of this release.

File: assets/js/admin.js
Comment (mikeyarce) [reply assets/js/admin.js]: 😂   Sometimes I write things in a codepen or something and then move it over and forget to check for this stuff...

File: includes/admin/class-wp-job-manager-cpt.php
Comment (renatho) [reply includes/admin/class-wp-job-manager-cpt.php:580]: > What do you think could happen in this case?

I thought about weird things like having the links, and buttons as icons mixed. But maybe it's not a big deal. I'd just confirm if our official extensions aren't using it. And if some is using it, we could also update the HTML, if needed. :)

File: includes/admin/class-wp-job-manager-promoted-jobs.php
Comment (mikeyarce) [reply includes/admin/class-wp-job-manager-promoted-jobs.php]: So one weird caveat I found is that you can't have a `dialog` inside of a `template` - because of how the Shadow DOM works and how components are mounted.  So we need to have a fallback in `get_promote_jobs_template` that has the dialog in it.  Just wanted to make a note if it here!

File: includes/admin/class-wp-job-manager-promoted-jobs.php
Comment (mikeyarce) [reply includes/admin/class-wp-job-manager-promoted-jobs.php]: I like that idea!

File: includes/admin/class-wp-job-manager-promoted-jobs.php
Comment (mikeyarce) [reply includes/admin/class-wp-job-manager-promoted-jobs.php]: Do we need to add `i18n` functions since it's being generated from our side?

File: assets/js/admin.js
Comment (mikeyarce) [reply assets/js/admin.js]: I'm not sure I totally understand what you mean here - do you mind explaining a bit more? 😄

File: assets/css/admin.scss
Comment (renatho) [inline assets/css/admin.scss]: In this one, since it's a class, we could use the format `.promote-jobs-dialog`. ;)

File: includes/admin/class-wp-job-manager-promoted-jobs.php
Comment (renatho) [inline includes/admin/class-wp-job-manager-promoted-jobs.php]: We need to also update the classes here for the fallback.

File: assets/css/admin.scss
Comment (renatho) [inline assets/css/admin.scss]: Another thing I was thinking here is that it would be interesting to give more CSS classes to the elements. Usually, it's a good practice, but mainly here because we're doing it a little more dynamic getting the content from wpjobmanager.com. So if we decide to add any other image to any part, for example, it would have a `width: 100%`.

The same would apply to the button, list, and so on.

File: includes/admin/class-wp-job-manager-promoted-jobs.php
Comment (renatho) [reply includes/admin/class-wp-job-manager-promoted-jobs.php]: Yep. Since it's the fallback content, probably it's worth having it translated.

File: assets/js/admin.js
Comment (renatho) [reply assets/js/admin.js]: Sorry! Sure!

I mean that we could separate a little more the classes and the functions of the generic part of the modal, and add with other classes and functions the specific details of the promote button.

An application example: if we decide later to create a confirmation modal for the job deletion, we could just use the same modal, applying the extra logic only for the deletion.

In practice, it would mean using a more generic class to the dialog, (maybe something like `wpjm-modal`?). And separate functions to open/close the modal (the generic part) and update the `href` button (the specific promote button part).

As a final result, we'd be able to easily create other modals.

I think it would also help with the deactivate modal logic.

Does it make sense to you?

File: assets/css/admin.scss
Comment (renatho) [reply assets/css/admin.scss]: I'm marking it as unsolved because it still seems to miss some spacing based on the design, okay?

Comment (mikeyarce) [reply assets/css/admin.scss]: OH I see what you mean.  I thought you meant the Checkmark and the Item, but you mean each item!  
Fixed in 3973a8885bb9b4139d5bebe92c92b8d1d94ad93b

File: assets/js/admin.js
Comment (mikeyarce) [reply assets/js/admin.js]: Makes sense!  Fixed in 11d6fd03e117f5b3ccc2d208b168c8308f2f6e6f

File: includes/admin/class-wp-job-manager-cpt.php
Comment (mikeyarce) [reply includes/admin/class-wp-job-manager-cpt.php:580]: The idea I had last night:  Once this is merged, I am going to ping our WPJM Friends in Slack and let them know about this coming change and see if they have questions or concerns.  How does that sound?

File: includes/admin/class-wp-job-manager-promoted-jobs.php
Comment (mikeyarce) [reply includes/admin/class-wp-job-manager-promoted-jobs.php]: Fixed in 25af99885951b8916910f87b96c62472bb50f77e

File: assets/css/admin.scss
Comment (mikeyarce) [reply assets/css/admin.scss]: Fixed in acafdbb6aee91bd404fd30cfc7e34803509c4389

File: includes/admin/class-wp-job-manager-promoted-jobs.php
Comment (mikeyarce) [reply includes/admin/class-wp-job-manager-promoted-jobs.php]: The template doesn't include the dialog - so I don't think we do in this case?  I found out that you can't have dialogs in templates, they just don't seem to work.

File: assets/css/admin.scss
Comment (mikeyarce) [reply assets/css/admin.scss]: Done in f4cad966c8698eddfaa04a6b5425e2a08e32d047

File: includes/admin/class-wp-job-manager-cpt.php
Comment (renatho) [reply includes/admin/class-wp-job-manager-cpt.php:580]: Not sure if it's so important in order to make some type of "noise". I think it would be more important for us to check the plugins we maintain for now. I took the opportunity to run a search all in the plugins here for `job_manager_admin_actions`, and I didn't find anything. So I think we're fine here. :)

File: includes/admin/class-wp-job-manager-promoted-jobs.php
Comment (renatho) [reply includes/admin/class-wp-job-manager-promoted-jobs.php]: Ah, I meant about classes like the `class="column-left"` that we updated.

Comment (mikeyarce) [reply includes/admin/class-wp-job-manager-promoted-jobs.php]: Gotcha!  Done in a44c9bcd56323db227551d8426c1467637876878

**Comment by renatho** [comment]
@mikeyarce, @yscik brought an idea in today's meeting to also bring the styles from wpjobmanager.com. Do you think we could easily extract the styles related to the modal content? If not, we could create a separate issue to do it later.

I'd suggest also adding it through `get_promote_jobs_template` in a `<style>` tag until we fetch it.

File: includes/admin/class-wp-job-manager-promoted-jobs.php
Comment (renatho) [inline includes/admin/class-wp-job-manager-promoted-jobs.php]: Just a final thing and I think we're ready to merge it.

We also need to fix the `promoteDialog` id to `promote-dialog`.

And for the `promote-jobs-dialog` class, in that idea to make it a little more generic to be able to re-use in the future in other contexts, we could call it as `wpjm-dialog`.

**Review by renatho — COMMENTED** [review]
Hey @mikeyarce! Thank you for all the tweaks.
I sent a last thing and I think we're ready to merge after that.

Also, about my previous comment, we can open a separate issue in order to merge it earlier. 😉

File: includes/admin/class-wp-job-manager-promoted-jobs.php
Comment (mikeyarce) [reply includes/admin/class-wp-job-manager-promoted-jobs.php]: Done in cbe1c8ebe577384a0a4f9b3ec53817e425895ba5

**Comment by mikeyarce** [comment]
> Do you think we could easily extract the styles related to the modal content? If not, we could create a separate issue to do it later.
> I'd suggest also adding it through `get_promote_jobs_template` in a `<style>` tag until we fetch it.

Yeah the styles are easy to extract - and the nice thing about using `<style>` in a template is that it's scoped only to that template so no need to worry about overriding other styles.

I think a new issue for this would be good 👍🏽

**Comment by renatho** [comment]
> Yeah the styles are easy to extract - and the nice thing about using <style> in a template is that it's scoped only to that template so no need to worry about overriding other styles.

TIL! That's great!

**Review by renatho — APPROVED** [review]
Thank you for the changes @mikeyarce!

I added to [this issue](https://github.com/Automattic/WP-Job-Manager/issues/2464) the part of extracting the styles.

**Comment by yscik** [comment]
Hey @mikeyarce, sorry for the late review, but I think there is some misunderstanding with the template stuff.

The role of the template tag and the custom element would be a means of filling in the row-specific items, on the client side, setting the main button to point to the right URL. Ideally it would work something like this:

- In PHP, get the content of the modal from wpjobmanager.com (or cache)
- In PHP, output it in PHP inside a `<template>` tag
- In JS, define a custom element from the template tag's content
- In JS, when clicking the Promote button, create a new instance of that custom element, and fill the button slot with the proper URL for that row. Something like:

```js
dialog.innerHTML = `<promote-job-modal>
	<div slot="buttons">
		<a href="...?postId=${postId}" class="...">Promote your job</button>
		...
	</div>
<promote-job-modal>`
```

PHP doesn't need to output an instance of `<promote-job-modal>`, we would be creating those dynamically when showing a modal. 

I don't think we need any fixed structure for the content of the modal, and not sure that having a template in the plugin that the remote content fills adds any value. The modal content can be a freeform HTML, at some point even an editable post in the remote website, that right now only needs a single slot defined for the dynamic link.

**Pull request merged**
